util/gem5-resources-manager-v2/src/gem5_resources_manager/data_source/atlas_data_source.py

```diff
@@ -0,0 +1,170 @@
+from .abstract_data_source import AbstractDataSource
+from .resource_category import ResourceCategory
+from .exception import Gem5DataSourceSchemaViolation, Gem5DataSourceEntryNotFound
+from ..schemas.json_validator import JSONValidator
+from pymongo.errors import ConnectionFailure, ConfigurationError
+from pymongo import MongoClient
+from typing import Any, Dict, Optional
+from bson import json_util
+import json
+import os
+
+
+class AtlasDataSource(AbstractDataSource):
+    def __init__(self, mongo_uri, database_name, collection_name) -> None:
+        super().__init__()
+        self.mongo_uri = mongo_uri
+        self.database_name = database_name
+        self.collection_name = collection_name
+        self.validator = JSONValidator(
+            "https://resources.gem5.org/gem5-resources-schema.json"
+        )
+
+    def __enter__(self):
+        self.open()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        return super().__exit__(exc_type, exc_value, traceback)
+
+    def open(self):
+        try:
+            client = MongoClient(self.mongo_uri)
+            client.admin.command("ping")
+        except ConnectionFailure:
+            client.close()
+            raise ConnectionFailure("Failed to connect to MongoDB")
+        except ConfigurationError:
+            client.close()
+            raise ConfigurationError("Invalid MongoDB URI")
+
+        database = client[self.database_name]
+        if database.name not in client.list_database_names():
+            client.close()
+            raise ConnectionFailure("Database does not exist")
+        collection = database[self.collection_name]
+        if collection.name not in database.list_collection_names():
+            client.close()
+            raise ConnectionFailure("Collection does not exist")
+
+        self.collection = collection
+
+    def close(self):
+        pass
+
+    def resource_exists(
+        self, resource_id: str, resource_version: Optional[str] = "1.0.0"
+    ) -> bool:
+        resource = (
+            self.collection.find(
+                {
+                    "id": resource_id,
+                    "resource_version": resource_version,
+                },
+                {"_id": 0},
+            )
+            .sort("resource_version", -1)
+            .limit(1)
+        )
+        json_resource = json_util.dumps(resource)
+        res = json.loads(json_resource)
+        if res == []:
+            return False
+        return True
+
+    def get_resource_category(
+        self, resource_id: str, resource_version: str
+    ) -> ResourceCategory:
+        resource = (
+            self.collection.find(
+                {
+                    "id": resource_id,
+                    "resource_version": resource_version,
+                },
+                {"_id": 0},
+            )
```

Comment:
You're writing this over and over. Make a private function calkled "get_resource" so you just write it the once.